### PR TITLE
I LOVE CIRCLES (heartbeat widget)

### PR DESCRIPTION
**Pull request by InvincibleRMC**
# Pull Request

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Add Circles to the gui.

## Related Tickets & Documents
#71 

- Related Issue #
- Closes #
#107

### CODE REVIEW

**Comment by InvincibleRMC** [comment]
@NoahMollerstuen Your custom styles don't seem to work when running the guis as a node.

**Comment by NoahMollerstuen** [comment]
> @NoahMollerstuen Your custom styles don't seem to work when running the guis as a node.

What do you mean by that? Styles aren't applied at all, the custom styles aren't applied (but you get dark theme?) or you just can't change the theme? How exactly are you starting the node?

**Comment by InvincibleRMC** [comment]
When I use ros2 run gui run_pilot or run_operator it runs the guis in lightmode but no custom coloration.

File: src/surface/gui/gui/widgets/circle.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+from typing import Callable, Optional, TypeAlias
+
+from gui.styles.custom_styles import WidgetState
+from PyQt6.QtCore import QSize, Qt
+from PyQt6.QtGui import QColor
+from PyQt6.QtWidgets import QPushButton, QWidget
+
+QColorConstantsAlias: TypeAlias = QColor | Qt.GlobalColor | int
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/circle.py:8]: Comment this crap

File: src/surface/gui/gui/widgets/circle.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+
+class CircleButton(QPushButton):
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/circle.py:11]: Investigate using QWidget to do this instead

**Review by benjaminwp18 — CHANGES_REQUESTED** [review]

File: src/surface/gui/gui/widgets/circle.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        self.setStyleSheet(f"{stylesheet}border-radius: {radius}px;")
+
+    def set_color(self, color: QColor | Qt.GlobalColor) -> None:
+        if isinstance(color, Qt.GlobalColor):
+            color = QColor(color)
+        style = f"background-color: rgb({color.red()}, {color.green()}, {color.blue()});"
+        self.setStyleSheet(f"{self.styleSheet()}{style}")
+
+
+class CircleButton(QPushButton, CircleInterface):
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/circle.py:23]: Delete this

File: src/surface/gui/gui/styles/custom_styles.py [hunk 4]
```diff
@@ -19,3 +22,30 @@ class WidgetState(Style):
     # some external factor, either another widget or something external to the gui
     # For example, a the arm button when the pi is not connected
     INACTIVE = "inactive"
+
+    # Removes any state
+    NO_STATE = ""
+
+
+class WidgetStateInterface(QWidget):
```
Comment (benjaminwp18) [inline src/surface/gui/gui/styles/custom_styles.py:30]: Rename to something like `StatefulWidget`, `HasWidgetState`, etc.

Comment (benjaminwp18) [reply src/surface/gui/gui/styles/custom_styles.py:30]: Use this everywhere we want statefulness
Move WidgetStyle to be "private" member of WidgetStateInterface

Comment (benjaminwp18) [reply src/surface/gui/gui/styles/custom_styles.py:30]: Put stateful button in here

Comment (benjaminwp18) [reply src/surface/gui/gui/styles/custom_styles.py:30]: Rename all the stateful stuff to Indicator-

File: src/surface/gui/gui/widgets/circle.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+from typing import Callable, Optional
+from gui.styles.custom_styles import WidgetStateInterface
+from PyQt6.QtCore import QSize, Qt
+from PyQt6.QtGui import QColor
+from PyQt6.QtWidgets import QPushButton, QWidget, QLabel
+
+
+class CircleInterface(QWidget):
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/circle.py:8]: Change to `Circle(QLabel)`

File: src/surface/gui/gui/widgets/circle.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                 radius: int = 50,
+                 color: Optional[QColor | Qt.GlobalColor] = None) -> None:
+        super().__init__(parent)
+        CircleInterface.__init__(self, parent, radius)
+
+        if color:
+            self.set_color(color)
+
+
+class Indicator(Circle, WidgetStateInterface):
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/circle.py:50]: `IndicatorCircle`

File: src/surface/gui/gui/widgets/heartbeat.py [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+
+        pi_status_layout = QHBoxLayout()
+        self.pi_indicator = QLabel('No Pi Status')
+        self.pi_indicator.setFont(font)
+        pi_status_layout.addWidget(self.pi_indicator)
+        self.pi_indicator_circle = Indicator(radius=10)
+        pi_status_layout.addWidget(self.pi_indicator_circle)
+        heartbeat_layout.addLayout(pi_status_layout)
+
+        pi_status_layout = QHBoxLayout()
```
Comment (benjaminwp18) [inline src/surface/gui/gui/widgets/heartbeat.py:34]: Rename to `pixhawk_status_layout`

**Review by benjaminwp18 — CHANGES_REQUESTED** [review]

**Review by benjaminwp18 — APPROVED** [review]

**Pull request merged**